python/tank/bootstrap/manager.py

```diff
@@ -794,24 +814,29 @@ def _cache_apps(self, pipeline_configuration, config_engine_name, progress_callb
         else:
             raise TankBootstrapError("Unsupported caching_policy setting %s" % self._caching_policy)
 
+        descriptors = {}
         # pass 1 - populate list of all descriptors
         for env_name in pipeline_configuration.get_environments():
             env_obj = pipeline_configuration.get_environment(env_name)
             for engine in env_obj.get_engines():
                 if engine_constraint is None or engine == engine_constraint:
-                    descriptors.append(env_obj.get_engine_descriptor(engine))
+                    descriptor = env_obj.get_engine_descriptor(engine)
+                    descriptors[descriptor.get_uri()] = descriptor
```

Comment:
nice! was looking at the code the other day and was thinking that we should fix this :)